packages/@react-stately/numberfield/src/useNumberFieldState.ts

```diff
@@ -11,159 +11,240 @@
+    }
+    // if it failed to parse, then reset input to formatted version of current number
+    if (isNaN(currentlyParsed)) {
+      setInputValue(inputValueFormatter.format(numberValue));
+      return;
+    }
+
+    let clampedValue;
+    if (!isNaN(currentlyParsed)) {
+      clampedValue = clamp(currentlyParsed, minValue, maxValue, step);
+    } else {
+      clampedValue = clamp(numberValue, minValue, maxValue, step);
+    }
+    clampedValue = numberParser.round(clampedValue);
+    let result = isNaN(clampedValue) ? '' : inputValueFormatter.format(clampedValue);
+    setNumberValue(clampedValue);
+    // in a controlled state, the numberValue won't change, so we won't go back to our old input without help
+    if (value === undefined) {
+      setInputValue(result);
+    } else {
+      setInputValue(inputValueFormatter.format(numberValue));
+    }
+  };
+
+  let increment = useCallback((isChained: boolean = false) => {
     setNumberValue((previousValue) => {
+      let prev = !isChained ? currentlyParsed : previousValue;
+      if (isNaN(prev)) {
+        // if the input is empty, start from 0
+        prev = 0;
+        if (!isNaN(props.minValue) && prev < props.minValue) {
+          // unless zero is less than the min value, then start well below it so we clamp to the min
+          prev = -Infinity;
+        }
+      }
+
+      let clampStep = !isNaN(step) ? step : 1;
+      if (intlOptions.style === 'percent' && isNaN(step)) {
+        clampStep = 0.01;
+      }
+      let clampedValue = clamp(prev, minValue, maxValue, step);
+      if (clampedValue > prev) {
+        return clampedValue;
+      }
       const newValue = clamp(
-        handleDecimalOperation('+', previousValue, step),
+        handleDecimalOperation('+', prev, clampStep),
         minValue,
-        maxValue
+        maxValue,
+        step
       );
-
-      updateValidation(newValue);
-      setInputValue(inputValueFormatter.format(newValue));
       return newValue;
     });
-  };
-
-  let incrementToMax = useCallback(() => {
-    if (maxValue != null) {
-      setNumberValue(maxValue);
-      setInputValue(inputValueFormatter.format(maxValue));
-    }
-  }, [inputValueFormatter, maxValue, setNumberValue]);
+  }, [setNumberValue, minValue, maxValue, step, isMaxRange, intlOptions, currentlyParsed]);
 
-  let decrement = () => {
+  let decrement = useCallback((isChained: boolean = false) => {
     setNumberValue((previousValue) => {
+      let prev = !isChained ? currentlyParsed : previousValue;
+      // if the input is empty, start from the max value when decrementing
+      if (isNaN(prev)) {
+        prev = 0;
+        // unless zero is greater than the max value, then start well above it so we clamp to the max
+        if (!isNaN(props.maxValue) && prev > maxValue) {
+          prev = Infinity;
+        }
+      }
+      let clampStep = !isNaN(step) ? step : 1;
+      if (intlOptions.style === 'percent' && isNaN(step)) {
+        clampStep = 0.01;
```

Comment:
Maybe we should normalize the options in a single place? noticed this is duplicated between increment and decrement.